File: apps/mobile/androidApp/src/main/kotlin/co/heva/piro/android/alert/AlertDetailViewModel.kt
package co.heva.piro.android.alert

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import co.heva.piro.shared.api.PiroApiClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/** Loads one alert's detail and acknowledges it. */
class AlertDetailViewModel(
    private val api: PiroApiClient,
    private val alertId: Int,
) : ViewModel() {

    private val _state = MutableStateFlow(AlertDetailUiState())
    val state: StateFlow<AlertDetailUiState> = _state.asStateFlow()

    init {
        load()
    }

    fun load() {
        viewModelScope.launch {
            _state.update { it.copy(loading = true, error = null) }
            try {
                val alert = api.getAlert(alertId)
                _state.update { it.copy(loading = false, alert = alert) }
            } catch (e: Exception) {
                _state.update { it.copy(loading = false, error = "Could not load the alert.") }
            }
        }
    }

    fun acknowledge() {
        if (_state.value.acknowledging) return
        viewModelScope.launch {
            _state.update { it.copy(acknowledging = true, error = null) }
            try {
                val updated = api.acknowledgeAlert(alertId)
                _state.update { it.copy(acknowledging = false, alert = updated) }
            } catch (e: Exception) {
                _state.update { it.copy(acknowledging = false, error = "Acknowledge failed.") }
            }
        }
    }
}

File: apps/mobile/iosApp/Piro/Alerts/AlertDetailViewModel.swift
import Foundation
import Shared

/// Loads one alert's detail and acknowledges it — the SwiftUI counterpart of the Android
/// `AlertDetailViewModel`.
@MainActor
final class AlertDetailViewModel: ObservableObject {
    @Published var loading = true
    @Published var alert: AlertDetail?
    @Published var error: String?
    @Published var acknowledging = false

    private let api: PiroApiClient
    private let alertId: Int32

    init(api: PiroApiClient, alertId: Int) {
        self.api = api
        self.alertId = Int32(alertId)
    }

    func load() async {
        loading = true
        error = nil
        do {
            alert = try await api.getAlert(id: alertId)
        } catch {
            self.error = "Could not load the alert."
        }
        loading = false
    }

    func acknowledge() async {
        guard !acknowledging else { return }
        acknowledging = true
        error = nil
        do {
            alert = try await api.acknowledgeAlert(id: alertId)
        } catch {
            self.error = "Acknowledge failed."
        }
        acknowledging = false
    }
}
